pytissueoptics/rayscattering/opencl/CLPhotons.py

```diff
@@ -39,27 +40,74 @@ def setContext(self, scene: RayScatteringScene, environment: Environment, logger
         self._materials = [worldMaterial]
         self._sceneLogger = logger
 
-        safetyFactor = 1.8
-        avgInteractions = int(-np.log(self._weightThreshold) / worldMaterial.getAlbedo())
-        self._requiredLoggerSize = self._N * int(safetyFactor * avgInteractions)
-
     def propagate(self):
-        photons = PhotonCL(self._positions, self._directions)
-        materials = MaterialCL(self._materials)
-        logger = DataPointCL(size=self._requiredLoggerSize)
-        randomNumbers = RandomNumberCL(size=self._N)
-        seeds = SeedCL(size=self._N)
-
         program = CLProgram(sourcePath=PROPAGATION_SOURCE_PATH)
+        params = CLParameters()
+
+        if params.photonAmount >= self._N:
+            params.photonAmount = self._N
+
+        kernelPhotons = PhotonCL(self._positions[0:params.photonAmount], self._directions[0:params.photonAmount])
+        photonPool = PhotonCL(self._positions[params.photonAmount:], self._directions[params.photonAmount:])
+        photonPool.make(program.device)
+        materials = MaterialCL(self._materials)
+        seeds = SeedCL(params.photonAmount)
 
+        photonCount = 0
+        batchCount = 0
+        logArrays = []
         t0 = time.time_ns()
-        maxInteractions = np.uint32(self._requiredLoggerSize // self._N)
-        program.launchKernel(kernelName='propagate', N=self._N,
-                             arguments=[self._N, maxInteractions, self._weightThreshold,
-                                        photons, materials, logger, randomNumbers, seeds])
-        log = program.getData(logger)
-        t1 = time.time_ns()
-        print("CLPhotons.propagate: {} s".format((t1 - t0) / 1e9))
 
+        while photonCount < self._N:
+            logger = DataPointCL(size=params.maxLoggableInteractions)
```

Comment:
Shouldn't we be using the autoReset behavior instead of creating a new logger object?

Reply:
yes

Reply:
but, the size of the logger might vary.

Reply:
I guess for now it isnt too bad if the last batch is bigger than necessary